### PR TITLE
etcd, clustermesh: generalize and untangle the custom dialer logic for automatic DNS name to service ClusterIP translation 

### PR DESCRIPTION
**Pull request by giorio94**
Please review commit by commit, and refer to the individual messages for additional details.

At a high level, this PR introduces a new generic context dialer with support for configuring custom resolvers, as well as implements a service resolver to map DNS names matching Kubernetes services to the corresponding ClusterIP address. They generalize and replace already existing logic used for etcd and clustermesh to avoid depending on an external DNS server (e.g., CoreDNS) from host-netns pods. Overall, the goal is to simplify the overall logic, make it possibly suitable for additional use-cases, and untangle it from the ServiceCache concept, which is not always available, and led to the introduction of additional hacks. 

```release-note
etcd, clustermesh: generalize and untangle the custom dialer logic for automatic DNS name to service ClusterIP translation
```

### CODE REVIEW

**Review by aditighag — COMMENTED** [review]
>It's main purpose is to be used
as a DialOption for etcd, and resolve DNS names representing k8s
services to the corresponding ClusterIP without depending on CoreDNS.

If I understand correctly, the main highlight of this PR is to expose a k8s service resolver that resolves service names to clusterIPs via K8s api store resource instead of probing core DNS pods. Is that right?
Maybe I'm missing context around etcd/clustermesh, but I have two main high level questions - 
(1) Why is probing kube-api server considered more reliable than core dns pods?
(2) What happens when the agent-kube api server connectivity is down? Should the new service dialer fallback to coredns resolver?

**Comment by giorio94** [comment]
> If I understand correctly, the main highlight of this PR is to expose a k8s service resolver that resolves service names to clusterIPs via K8s api store resource instead of probing core DNS pods. Is that right?

That's correct. The same logic is already present today in Cilium, and this PR refactors it into an independent module, decoupled from the `ServiceCache` abstraction (as that's only relevant inside the agent), which can be consumed when constructing the different etcd clients (although it could be valuable for other use-cases as well). The approach is also generalized for easier extension in the future, e.g., introducing new resolvers for clustermesh-specific use-cases.

> (1) Why is probing kube-api server considered more reliable than core dns pods?

The core reason for this translation logic is that by default pods running in the host-network namespace (both Cilium Agents and the Cilium Operator in this case) target the DNS server configured in the host, and not the cluster DNS server. Hence, they are not able to resolve service names, unless they are configured with [the `ClusterFirstWithHostNet` DNS policy](https://kubernetes.io/docs/concepts/services-networking/dns-pod-service/#pod-s-dns-policy). In other words, this logic avoids needing to tune that value, and prevents depending on CoreDNS to be up and running to perform the translation (e.g., at bootstrap time).

> (2) What happens when the agent-kube api server connectivity is down? Should the new service dialer fallback to coredns resolver?

If the resolver fails to find a mapping, then the dialer will simply use the original address, falling back to normal DNS translation. The only exception is during the initial synchronization of the store, as requests matching a service name are blocked until the service store has synchronized (similarly to what happens with the current implementation).

**Review by aditighag — APPROVED** [review]
Thanks for the context. Looks good for my CODEOWNERS.

> Hence, they are not able to resolve service names, unless they are configured with [the `ClusterFirstWithHostNet` DNS policy](https://kubernetes.io/docs/concepts/services-networking/dns-pod-service/#pod-s-dns-policy).

That's what cilium-agent and operator DS are configured with for the managed etcd mode - https://github.com/cilium/cilium/blob/main/install/kubernetes/cilium/templates/cilium-operator/deployment.yaml#L262-L262 and https://github.com/cilium/cilium/blob/main/install/kubernetes/cilium/templates/cilium-agent/daemonset.yaml#L762.

**Review by derailed — APPROVED** [review]
@giorio94 Nice clean up. Good work Marco!

File: pkg/dial/dialer.go
Comment (squeed) [inline pkg/dial/dialer.go:17]: Does it make sense to define an explicit "not found" error type, to distinguish between empty lookup and transient failure?

**Review by squeed — APPROVED** [review]
really nice cleanup!

**Comment by giorio94** [comment]
> That's what cilium-agent and operator DS are configured with for the managed etcd mode

Yep, it is not exactly clear to me the reason for setting both the DNS policy and having the resolver in that configuration (FWIW, the managed mode is also being removed in https://github.com/cilium/cilium/pull/32921). However, my understanding is that etcd could be behind a service also in other cases (besides the KVStoreMesh case), and in general the resolver shouldn't do any harm (it is supposed to be transparent if the URL does not match a service).

File: pkg/dial/dialer.go
Comment (giorio94) [reply pkg/dial/dialer.go:17]: Hmm, I'm not actually that sure whether there's a clear distinction between temporary and not found errors, in the sense that even a not found error may be temporary (e.g., the creation event for the given service has not yet been received). Considering that the error possibly returned by a resolver is ignored by the dialer anyway to keep the behavior as transparent as possible (hence just dialing with the original hostname), there's probably little value in distinguishing the type of error.

**Comment by giorio94** [comment]
Rebased onto main to fix conflict.

File: daemon/cmd/daemon_main.go
Comment (marseel) [inline daemon/cmd/daemon_main.go:1588]: Shouldn't we check if resolver is not nil?
Seems to me like it will cause nil pointer dereference in `newContextDialer`

**Review by marseel — CHANGES_REQUESTED** [review]
Clustermesh lgtm, kvstore seems to have a small bug?

File: daemon/cmd/daemon_main.go
Comment (giorio94) [reply daemon/cmd/daemon_main.go:1588]: In the current implementation, the resolver is never expected to be nil, as retrieved from hive, and always [provided by the dedicated cell](https://github.com/giorio94/cilium/blob/mio%2Fcustom-dialer/daemon/cmd/cells.go#L248).

**Review by marseel — APPROVED** [review]
thanks for explanation, lgtm

**Comment by giorio94** [comment]
@lambdanis Gentle ping :pray:

**Review by ghost — APPROVED** [review]
docs good

**Comment by giorio94** [comment]
Rebased onto main to fix conflict.

**Comment by giorio94** [comment]
Rebased onto main to fix conflict.

**Pull request merged**
